### DISCUSSION TITLE
Improving the usage of the Lua files

### DISCUSSION THREAD

**Discussion by LeonardoFeriato** [zhad3/zrenderer#45] [Ideas]
Hi man.

I suggest optimizing the usage of the Lua files in the project if possible.
The app was built using the logic inside kRO lua files, however, in other servers like bRO and iRO this logic doesn't work with all items.

My suggestion is to be able to edit lua files of kRO, like <code>data\luafiles514\lua files\datainfo\spriterobeid.*</code>, <code>data\luafiles514\lua files\datainfo\weapontable.*</code> OR if the <code>resolver.d</code> file doesn't find the sprite, use another editable TXT file in <code>resolver_data</code> folder to create a new map to the right sprite.

To illustrate with this item https://www.divine-pride.net/database/item/480407/
It doesn't exist in kRO, but we know DBName and NumClass (sprite id).
Is there a way to use this data to force render the item?

**Comment by zhad3** [comment]
Hi @LeonardoFeriato, the `resolver_data` folder is for data that is hardcoded in the client. **Do you know from where bRO or iRO get their information?** Technically it is possible to add "custom fallback data" to read from but it kinda eliminates the idea of the service *just working*. I would first try to mimic the official client unless of course they have all this data hardcoded and don't use any other lua or txt file from the grf.

**Reply by zhad3** [reply]
>  It doesn't exist in kRO, but we know DBName and NumClass (sprite id).

What zrenderer requires is a "NumClass (sprite id)" to filename mapping. This mapping exists (in kRO) in the file `accname.lub`.
For example:
```
# accessoryid.lub (essentially the numclass or sprite id)
ACCESSORY_C_Frog_Hood = 2010

# accname.lub
[ACCESSORY_IDs.ACCESSORY_C_Frog_Hood] = "_C_Frog_Hood",
```
It will then look up the sprite with filename `_C_Frog_Hood` in the grf.

**Reply by LeonardoFeriato** [reply]
Getting this data is quite easy, maybe using GRF Editor or something else.
Is it possible to create a "fallback data" file like this for the project?

```
# acessory_fallback.txt
2010 = "_C_Frog_Hood",
```

**Comment by LeonardoFeriato** [comment]
Hi @zhad3,

> Do you know from where bRO or iRO get their information?

I have no idea, I try to find out without success.

**Reply by zhad3** [reply]
I downloaded and updated bRO but I couldn't find https://www.divine-pride.net/database/item/480407/ in the iteminfo file. Do you know in which version it is available?

I found this bRO exclusive item https://www.divine-pride.net/database/item/400476/. Checking the files though they all seem to be in place as usual.

I might have misunderstood the idea here. The issue is not that they aren't found but rather you wish to merge different versions (kRO, jRO, bRO, iRO) together into one?

**Reply by LeonardoFeriato** [reply]
Oops, my bad.
 https://www.divine-pride.net/database/item/480407/ is in **iRO**
 https://www.divine-pride.net/database/item/20597 is in **bRO**

**Reply by LeonardoFeriato** [reply]
> I might have misunderstood the idea here. The issue is not that they aren't found but rather you wish to merge different versions (kRO, jRO, bRO, iRO) together into one?

Yeah, it's correctly. Because using only bRO files or iRO files I always have this issue https://github.com/zhad3/zrenderer/issues/36

**Reply by zhad3** [reply]
First to address the issue of the "shadowtable error" I have released a new version:https://github.com/zhad3/zrenderer/releases/tag/v1.4.1. Tested this locally against my up-to-date bRO and I was at least able to start it and render some sprites.

Can you give it a try with your bRO or iRO version?

**Reply by zhad3** [reply]
On the other hand I'll have to think how to make zrenderer version agnostic (allow simultaneous bRO, iRO, kRO, etc.)

**Comment by LeonardoFeriato** [comment]
Hey there, @zhad3 

I built the last version and to some requests using the bRO file and got this output:
```
[vibe-worker-7(tCKB) INF] While loading lua resource "data/luafiles514/lua files/datainfo/shadowtable.lub": D:\MA\shadowtable.lua:796: table index is nil
[main(Np7W) INF] 127.0.0.1:61313 - - 2024-Jul-07 15:26:34.1609136Z "POST /render?downloadimage&accesstoken=*** HTTP/1.1" 200 5594 "-" "PostmanRuntime/7.39.0" "Auto-generated Admin Token"
```

Is that what you looking for?

Using:
https://www.divine-pride.net/database/item/20597 is in bRO

Body:
```
{
  "gender": 0,
  "job": [
    "23"
  ],
  "head": 21,
  "headPalette": 1,
  "headdir": 0,
  "headgear": [
    0,
    0,
    0
  ],
  "garment": 79,
  "bodyPalette": 0,
  "action": 0,
  "canvas": "200x200+100+150",
  "outfit": 0
}
```

**Reply by zhad3** [reply]
Yes. Unlike before the service won't stop if the shadowtable cannot be loaded and you can continue to render sprites.
